File: source/Classical/Caesar.cs
using System.Text;
namespace Ciphers
{
    namespace Classical
    {
        public class Caesar : ClassicalCipher
        {
            public Caesar(string shift = "13")
            {
                _key = shift;

            }
            public override string Encode(string plain)
            {
                // Getting shift from the key field
                int shift = int.Parse(_key);

                //turning the string uppercase
                plain = plain.ToUpper();

                // Creating a stringbuilder entitiy to host the string
                StringBuilder cipherText = new StringBuilder();

                //looping through the message to get individual characters
                foreach (char character in plain.ToCharArray())
                {
                    if (!char.IsLetter(character))
                    {
                        cipherText.Append(character);
                        continue;
                    }
                    var pos = Array.IndexOf(_alphabet, character);
                    cipherText.Append(_alphabet[(pos - shift + _alphabet.Length) % _alphabet.Length]); ;
                }
                return cipherText.ToString();
            }
            public override string Decode(string cipher)
            {
                // Getting shift from the key field
                int shift = int.Parse(_key);
                //turning the string uppercase for consistency
                cipher = cipher.ToUpper();
                // Decomposing the alphabet into a char array
                // Creating a stringbuilder entitiy to host the string
                StringBuilder plainText = new StringBuilder();

                //looping through the message to get individual characters
                foreach (char character in cipher.ToCharArray())
                {
                    if (!char.IsLetter(character))
                    {
                        plainText.Append(character);
                        continue;
                    }
                    var pos = Array.IndexOf(_alphabet, character);
                    plainText.Append(_alphabet[(pos - shift + _alphabet.Length) % _alphabet.Length]);
                }
                return plainText.ToString().ToLower();
            }
        }
    }
}

File: src/ciphers/classical/caesar.go
package classical

import (
	"strconv"
	"strings"
	"unicode"
)

type Caesar struct {
	alphabet string
	key      string
}

func NewCaesar(alphabet string, key string) *Caesar {
	return &Caesar{alphabet: alphabet, key: key}
}

func (c Caesar) GetType() string {
	return "Caesar's cipher"
}

func (c Caesar) Encode(plainText string) string {
	// setup
	shift, _ := strconv.ParseInt(c.key, 10, 64)
	shift %= int64(len(c.alphabet))
	plainText = strings.ToUpper(plainText)
	var res strings.Builder
	var newIndex int
	//Encryption
	for _, char := range plainText {
		if !unicode.IsLetter(char) {
			res.WriteRune(char)
			continue
		}
		newIndex = strings.Index(c.alphabet, string(char)) + int(shift) + len(c.alphabet)
		newIndex %= len(c.alphabet)
		res.WriteRune(rune(c.alphabet[newIndex]))
	}
	return res.String()
}

func (c Caesar) Decode(cipherText string) string {
	// setup
	shift, _ := strconv.ParseInt(c.key, 10, 64)
	shift %= int64(len(c.alphabet))
	cipherText = strings.ToUpper(cipherText)
	var res strings.Builder
	var newIndex int
	//Encryption
	for _, char := range cipherText {
		if !unicode.IsLetter(char) {
			res.WriteRune(char)
			continue
		}
		newIndex = strings.Index(c.alphabet, string(char)) - int(shift) + len(c.alphabet)
		newIndex %= len(c.alphabet)
		res.WriteRune(rune(c.alphabet[newIndex]))
	}
	return strings.ToLower(res.String())
}
